### PR TITLE
libsmtutil

### LINKED ISSUE #8928: Extract libsmtutil out of libsolidity/formal
We should extract the contents of `libsolidity/formal` into a `libsmtutil` that can also be used by `libyul`. This should contain things that do not depend on anything Solidity related.
Potential cases for that:
- Sorts
- SolverInterface
- *Interface
- SMTPortfolio

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/8928

The CMake code that finds Z3 and CVC4 was moved up to the top level because it exports `HAVE_Z3` and `HAVE_CVC4` which are needed by both `libsmtutil` (since it has the solver specific interfaces) and `libsolidity/formal/*` (so that the includes work).

### CODE REVIEW

File: libsmtutil/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+endif()
+
+if (${CVC4_FOUND})
+  set(cvc4_SRCS CVC4Interface.cpp CVC4Interface.h)
+else()
+  set(cvc4_SRCS)
+endif()
+
+add_library(smtutil ${sources} ${z3_SRCS} ${cvc4_SRCS})
+target_link_libraries(smtutil PUBLIC langutil solutil Boost::boost)
```
Comment (chriseth) [inline libsmtutil/CMakeLists.txt]: Do you need langutil (scaner, parser)?

File: libsmtutil/Z3CHCInterface.h [hunk 2]
```diff
@@ -21,8 +21,8 @@
 
 #pragma once
 
-#include <libsolidity/formal/CHCSolverInterface.h>
-#include <libsolidity/formal/Z3Interface.h>
+#include <libsmtutil/CHCSolverInterface.h>
+#include <libsmtutil/Z3Interface.h>
 
 namespace solidity::frontend::smt
```
Comment (chriseth) [inline libsmtutil/Z3CHCInterface.h]: Should we also change the namespace? `solidity::smt`?

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -188,7 +188,7 @@ add_executable(soltest ${sources}
     ${libsolidity_util_sources}
     ${yul_phaser_sources}
 )
-target_link_libraries(soltest PRIVATE libsolc yul solidity yulInterpreter evmasm solutil Boost::boost Boost::filesystem Boost::program_options Boost::unit_test_framework evmc)
+target_link_libraries(soltest PRIVATE libsolc yul solidity yulInterpreter evmasm smtutil solutil Boost::boost Boost::filesystem Boost::program_options Boost::unit_test_framework evmc)
```
Comment (chriseth) [inline test/CMakeLists.txt]: Shouldn't most of these be pulled in via the dependency on solidity?

File: libsmtutil/CMakeLists.txt (hunk 1 above)
Comment (leonardoalt) [reply libsmtutil/CMakeLists.txt]: From `langutil` we need `Exceptions`, but might be header only

File: test/CMakeLists.txt (hunk 3 above)
Comment (leonardoalt) [reply test/CMakeLists.txt]: Maybe `smtutil`, `solutil` and `Boost::boost` yea

File: libsmtutil/Z3CHCInterface.h (hunk 2 above)
Comment (axic) [reply libsmtutil/Z3CHCInterface.h]: Yes please.

Comment (axic) [reply libsmtutil/Z3CHCInterface.h]: Or may be `solidity::smtutil`?

Comment (leonardoalt) [reply libsmtutil/Z3CHCInterface.h]: Going with `smtutil`

File: libsmtutil/CMakeLists.txt (hunk 1 above)
Comment (axic) [reply libsmtutil/CMakeLists.txt]: ~~What is needed from `solutil`?~~ `CommonIO`

File: libsmtutil/SolverInterface.h [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 #pragma once
 
-#include <libsolidity/formal/Sorts.h>
+#include <libsmtutil/Sorts.h>
 
 #include <libsolidity/ast/Types.h>
```
Comment (axic) [inline libsmtutil/SolverInterface.h]: Is this needed in this file?

File: libsmtutil/SolverInterface.h [hunk 5]
```diff
@@ -17,7 +17,7 @@
```
Comment (axic) [inline libsmtutil/SolverInterface.h:17]: Line 59 and 67 doesn't seem to be implemented in the z3/cvc4 files (instead `SymbolicTypes` has it), maybe it is some old leftover?

File: libsmtutil/SolverInterface.h (hunk 4 above)
Comment (leonardoalt) [reply libsmtutil/SolverInterface.h]: Yea, `Expression` has a constructor based on `Type`

File: libsmtutil/SolverInterface.h (hunk 5 above)
Comment (leonardoalt) [reply libsmtutil/SolverInterface.h:17]: Argh, true... `smtSort` is used exactly by the `Expression` constructor that takes `Type`

Comment (leonardoalt) [reply libsmtutil/SolverInterface.h:17]: Maybe I need to remove that and only take `Sort`, and the places that use `Expression(Type)` need to use `Expression(Sort)` instead

File: libsmtutil/CMakeLists.txt (hunk 1 above)
Comment (leonardoalt) [reply libsmtutil/CMakeLists.txt]: Done

**Comment by leonardoalt** [comment]
Updated.
Besides the review comments, also moved finding z3 and cvc4 to `cmake/EthDependencies`

File: libsolidity/formal/SMTEncoder.cpp [hunk 6]
```diff
@@ -1065,8 +1065,9 @@ void SMTEncoder::arrayIndexAssignment(Expression const& _expr, smtutil::Expressi
 		{
 			auto symbArray = dynamic_pointer_cast<smt::SymbolicArrayVariable>(m_context.expression(*base));
 			solAssert(symbArray, "");
+			auto baseType = base->annotation().type;
 			toStore = smtutil::Expression::tuple_constructor(
-				smtutil::Expression(base->annotation().type),
+				smtutil::Expression(make_shared<smtutil::SortSort>(smt::smtSort(*baseType)), baseType->toString(true)),
```
Comment (axic) [inline libsolidity/formal/SMTEncoder.cpp:1066]: Wait what, there were only 2 use cases for the separate constructor? :)

File: libsmtutil/Exceptions.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+#pragma once
+
+#include <libsolutil/Assertions.h>
+#include <libsolutil/Exceptions.h>
+
+namespace solidity::smtutil
+{
+
+struct SMTLogicError: virtual util::Exception {};
```
Comment (axic) [inline libsmtutil/Exceptions.h:26]: It is possible new exceptions may need to be added to StandardCompiler and/or CompilerStack and other places, but likely not. In the case of Yul exceptions there was one case where it was needed.

File: libsolidity/formal/SMTEncoder.cpp (hunk 6 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1066]: Yea, it was a new thing, came with the new tuple stuff

File: libsmtutil/Exceptions.h (hunk 7 above)
Comment (leonardoalt) [reply libsmtutil/Exceptions.h:26]: I don't think I get it, what would that be the case?
These occurrences use to be `solAssert`, they were just replaced.
The `solAssert` ones in `libsolidity/formal` stay the same.

Comment (axic) [reply libsmtutil/Exceptions.h:26]: There is a new exception type thrown. If you have code which excepts the old exception type that will break.

Comment (axic) [reply libsmtutil/Exceptions.h:26]: The mess I encountered was in #7884.

Comment (leonardoalt) [reply libsmtutil/Exceptions.h:26]: The old exception type is still thrown everywhere

**Comment by leonardoalt** [comment]
Rebased

**Comment by chriseth** [comment]
```
In file included from /root/project/libsolidity/formal/BMC.h:32,
                 from /root/project/libsolidity/formal/BMC.cpp:18:
/root/project/libsolidity/formal/SMTEncoder.h:142:51: error: 'solidity::frontend::smt::Expression' has not been declared
  142 |  void arrayPushPopAssign(Expression const& _expr, smt::Expression const& _array);
      |                                                   ^~~
```

**Comment by leonardoalt** [comment]
@chriseth Yep just fixed those

**Comment by chriseth** [comment]
Test failures.

**Review by chriseth — APPROVED** [review]

**Pull request merged**